data/src/main/java/org/cryptomator/data/cloud/crypto/CryptoImplDecorator.kt

```diff
@@ -332,16 +421,88 @@ abstract class CryptoImplDecorator(
 								)
 						}
 					}
+					thumbnailWriter.flush()
+					closeQuietly(thumbnailWriter)
 				}
 			} finally {
 				encryptedTmpFile.delete()
 				progressAware.onProgress(Progress.completed(DownloadState.decryption(cryptoFile)))
 			}
+
+			closeQuietly(thumbnailReader)
 		} catch (e: IOException) {
 			throw FatalBackendException(e)
 		}
 	}
 
+	private fun closeQuietly(closeable: Closeable) {
+		try {
+			closeable.close();
+		} catch (e: IOException) {
+			// ignore
+		}
+	}
+
+	private fun startThumbnailGeneratorThread(diskCache: DiskLruCache?, cacheKey: String, thumbnailReader: PipedInputStream): Future<*> {
+		return thumbnailExecutorService.submit {
+			try {
+				val options = BitmapFactory.Options()
+				val thumbnailBitmap: Bitmap?
+				options.inSampleSize = 4 // pixel number reduced by a factor of 1/16
+
+				val bitmap = BitmapFactory.decodeStream(thumbnailReader, null, options)
+				val thumbnailWidth = 100
+				val thumbnailHeight = 100
+				thumbnailBitmap = ThumbnailUtils.extractThumbnail(bitmap, thumbnailWidth, thumbnailHeight)
+
+				if (thumbnailBitmap != null) {
+					storeThumbnail(diskCache, cacheKey, thumbnailBitmap)
+				}
+
+				closeQuietly(thumbnailReader)
+			} catch (e: Exception) {
+				Timber.e("Bitmap generation crashed")
+			}
+		}
+	}
+
+	protected fun generateCacheKey(cloudFile: CloudFile): String {
+		return buildString {
+			if (cloudFile.cloud?.id() != null)
+				this.append(cloudFile.cloud!!.id())
+			else
+				this.append("c") // "common"
+			this.append("-")
+			this.append(cloudFile.path.hashCode())
+		}
+	}
+
+	private fun isGenerateThumbnailsEnabled(cache: DiskLruCache?, fileName: String): Boolean {
+		return sharedPreferencesHandler.useLruCache() &&
```

Comment:
From a user perspective, it is currently not obvious that you need to enable caching for thumbnail generation to work. We may need to show a dialogue when changing the thumbnail generation from NONE to something else, and also for when caching gets disabled.

Even better, we could also decouple it here completely from the response cache so that we do not depend on this response caching feature, right?